src/server/editorServices.ts

```diff
@@ -3692,6 +3694,26 @@ namespace ts.server {
             return files;
         }
 
+        /* @internal */
+        updateFileSystem(updatedFiles: protocol.FileSystemRequestArgs[] | undefined, deletedFiles?: string[]) {
+            Debug.assert(this.fshost instanceof VirtualFS.VirtualServerHost);
+            if (updatedFiles) {
+                for (const { file: path, fileContent: content } of updatedFiles) {
+                    if (this.fshost.fileExists(path)) {
+                        this.fshost.modifyFile(path, content);
+                    }
+                    else {
+                        this.fshost.ensureFileOrFolder({ path, content });
+                    }

```

Comment:
this,fshost.ensureFileOrFolder irrespective of fileExists?
And ensureFileOrFolder modifies the content of the file if file exists ?